zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+                            let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
+                            // concious decision here to not throw an error when the new amount is not higher
+                            // because the user would have to pay worst case fees then
+                            if amount > prev_highest_amount {
+                                let diff = amount.saturating_sub(prev_highest_amount);
+                                add_to_outcome_sum(diff);
+                                lock_info[i].1 = amount;
+                            }
+                        }
+                        Err(i) => {
+                            ensure!(
+                                lock_info.len() as u32 <= T::MaxGlobalDisputeVotes::get(),
+                                Error::<T>::MaxVotesReached
+                            );
+                            add_to_outcome_sum(amount);
+                            lock_info.insert(i, (market_id, amount));
+                        }
+                    }
+                    Ok(lock_info.len() as u32)
+                },
+            )?;
+
+            T::Currency::extend_lock(
+                T::VoteLockIdentifier::get(),
+                &sender,
+                amount,
+                WithdrawReasons::TRANSFER,
+            );
+
+            Self::deposit_event(Event::VotedOnOutcome(market_id, outcome, amount));
+            Ok(Some(T::WeightInfo::vote_on_outcome(outcome_owners_len, vote_lock_counter)).into())
+        }
+
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::unlock_vote_balance(
+            T::MaxGlobalDisputeVotes::get(),
+            T::MaxOwners::get()
+        ))]
+        pub fn unlock_vote_balance(
+            origin: OriginFor<T>,
+            voter: AccountIdLookupOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+            let voter = T::Lookup::lookup(voter)?;
+
+            let mut lock_needed: BalanceOf<T> = Zero::zero();
+            let vote_lock_counter =
+                <Locks<T>>::mutate(&voter, |LockInfo(ref mut lock_info)| -> u32 {
+                    let vote_lock_counter = lock_info.len() as u32;
+                    lock_info.retain(|&(market_id, locked_balance)| {
+                        // weight component MaxOwners comes from querying the winner information
+                        match <Winners<T>>::get(market_id) {
+                            Some(winner_info) if winner_info.is_finished => false,
+                            _ => {
+                                lock_needed = lock_needed.max(locked_balance);
+                                true
+                            }
+                        }
+                    });
+                    vote_lock_counter
+                });
+
+            if lock_needed.is_zero() {
+                T::Currency::remove_lock(T::VoteLockIdentifier::get(), &voter);
+            } else {
```

Comment:
I don't think this is correct. If the user has locked up some funds for votes, but the voting on those markets is not finished yet, the locks get cleared in the `Currency` pallet, but not in the locks and the votes still persist for that user.

Reply:
I am not able to get in your thoughts. Kindly asking, if you can read my comments and then ask your question again?

Reply:
Looking over it again I think it is correct. The `None` case in the `Winners` match arm should never happen, I think we don't necessarily need logs here though.